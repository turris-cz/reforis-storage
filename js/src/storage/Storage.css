/*
 * Copyright (c) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

table.table td div.form-group {
    margin-bottom: 0;
}
