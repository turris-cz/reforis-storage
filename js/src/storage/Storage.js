/*
 * Copyright (C) 2020-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import {
    API_STATE,
    ErrorMessage,
    Spinner,
    useAPIGet,
    formFieldsSize,
} from "foris";
import PropTypes from "prop-types";

import { PENDING_STORAGE_STATES } from "./constants";
import CurrentState from "./currentState/CurrentState";
import DrivesOperations from "./DrivesOperations";
import useStorageState from "./hooks";
import API_URLs from "../API";

Storage.propTypes = {
    ws: PropTypes.object.isRequired,
};

const HELP_TEXT = _(`
<p>Here you can set up where your persistent data should be stored.
If you want to use Nextcloud, LXC or other IO-intensive applications,
don't put them on internal flash, but always use external storage.
Also, make sure that your data will fit on the new drive before switching.</p>
<p>To set up external storage, you have to connect an external drive - like
an USB flash drive - and format it to Btrfs filesystem. All you have to do is
select the correct drive, the desired RAID level and click "Format & Set"
button in "Prepare storage" section. This will format the drive and will
ask you to reboot the router. During the reboot your /srv (directory
where all IO-intensive applications should reside) will get moved to this
new drive.</p>
<p>If you already have a device that you used with the Turris router before,
you can set it to be used as external storage in "Drive selection" section
without losing any data.</p>
`);

export default function Storage({ ws }) {
    const [getDrivesResponse, getDrives] = useAPIGet(API_URLs.drives);
    const [getSettingsResponse, getSettings] = useAPIGet(API_URLs.settings);
    const [storageState, getStorageState] = useStorageState(ws, getDrives);

    useEffect(() => {
        getDrives();
        getSettings();
    }, [getDrives, getSettings]);

    const updateUUIDCallback = () => {
        getDrives();
        getSettings();
        getStorageState();
    };

    let componentContent = null;
    if (
        storageState.state === API_STATE.ERROR ||
        getDrivesResponse.state === API_STATE.ERROR ||
        getSettingsResponse.state === API_STATE.ERROR
    ) {
        componentContent = <ErrorMessage />;
    } else if (
        storageState.state !== API_STATE.SUCCESS ||
        getDrivesResponse.state !== API_STATE.SUCCESS ||
        getSettingsResponse.state !== API_STATE.SUCCESS
    ) {
        componentContent = <Spinner />;
    } else {
        const storageIsPending =
            Object.keys(PENDING_STORAGE_STATES).includes(
                storageState.data.state
            ) || storageState.data.blocking;

        componentContent = (
            <div className={`${formFieldsSize}`}>
                <CurrentState
                    storageIsPending={storageIsPending}
                    {...storageState.data}
                    disk_mounted={getSettingsResponse.data.disk_mounted}
                />
                <DrivesOperations
                    drives={getDrivesResponse.data.drives}
                    currentUUID={storageState.data.uuid}
                    storageIsPending={storageIsPending}
                    updateUUIDCallback={updateUUIDCallback}
                />
            </div>
        );
    }

    return (
        <>
            <h1>{_("Storage")}</h1>
            <div dangerouslySetInnerHTML={{ __html: HELP_TEXT }} />
            {componentContent}
        </>
    );
}
