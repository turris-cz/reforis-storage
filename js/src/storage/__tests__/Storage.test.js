/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render } from "customTestRender";

import Storage from "../Storage";

describe("<Storage />", () => {
    it("should render component", () => {
        const { container, getByText } = render(<Storage />);
        expect(getByText("Storage")).toBeDefined();
    });
});
