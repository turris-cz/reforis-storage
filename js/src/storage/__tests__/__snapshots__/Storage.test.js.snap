// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Storage /> Should handle pending states. 1`] = `
<div>
  <h1>
    Storage
  </h1>
  <div>
    

    <p>
      Here you can set up where your persistent data should be stored.
If you want to use Nextcloud, LXC or other IO-intensive applications,
don't put them on internal flash, but always use external storage.
Also, make sure that your data will fit on the new drive before switching.
    </p>
    

    <p>
      To set up external storage, you have to connect an external drive - like
an USB flash drive - and format it to Btrfs filesystem. All you have to do is
select the correct drive, the desired RAID level and click "Format & Set"
button in "Prepare storage" section. This will format the drive and will
ask you to reboot the router. During the reboot your /srv (directory
where all IO-intensive applications should reside) will get moved to this
new drive.
    </p>
    

    <p>
      If you already have a device that you used with the Turris router before,
you can set it to be used as external storage in "Drive selection" section
without losing any data.
    </p>
    

  </div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Current State
    </h2>
    <div
      class="mb-3"
    >
      <div
        class="row no-gutters justify-content-center"
      >
        <div
          class="col-md-6 d-flex flex-column align-items-center justify-content-center"
        >
          <i
            class="fas fa-hdd fa-4x mb-1"
          />
          <h5>
            /dev/sdb
          </h5>
        </div>
        <div
          class="col-md-6"
        >
          <div
            class="card-body"
          >
            <div
              class="table-responsive"
            >
              <table
                class="table table-borderless table-hover"
              >
                <tbody>
                  <tr>
                    <th
                      scope="row"
                    >
                      State
                    </th>
                    <td>
                      <div
                        class="d-flex flex-row"
                      >
                        <div
                          class="spinner-border spinner-border-sm"
                          role="status"
                        >
                          <span
                            class="sr-only"
                          />
                        </div>
                        <div
                          class="spinner-text"
                        >
                          Â 
                          Formatting
                        </div>
                      </div>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Device
                    </th>
                    <td>
                      /dev/sdb
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      UUID
                    </th>
                    <td>
                      1234-1234-1234-1234-1234
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      RAID
                    </th>
                    <td>
                      single
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Storage /> Should render. 1`] = `
<div>
  <h1>
    Storage
  </h1>
  <div>
    

    <p>
      Here you can set up where your persistent data should be stored.
If you want to use Nextcloud, LXC or other IO-intensive applications,
don't put them on internal flash, but always use external storage.
Also, make sure that your data will fit on the new drive before switching.
    </p>
    

    <p>
      To set up external storage, you have to connect an external drive - like
an USB flash drive - and format it to Btrfs filesystem. All you have to do is
select the correct drive, the desired RAID level and click "Format & Set"
button in "Prepare storage" section. This will format the drive and will
ask you to reboot the router. During the reboot your /srv (directory
where all IO-intensive applications should reside) will get moved to this
new drive.
    </p>
    

    <p>
      If you already have a device that you used with the Turris router before,
you can set it to be used as external storage in "Drive selection" section
without losing any data.
    </p>
    

  </div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Current State
    </h2>
    <div
      class="alert alert-success"
    >
      
Device currently in use is

      <b>
        /dev/sdb
      </b>
       (UUID: 1234-1234-1234-1234-1234).
    </div>
    <div
      class="mb-3"
    >
      <div
        class="row no-gutters justify-content-center"
      >
        <div
          class="col-md-6 d-flex flex-column align-items-center justify-content-center"
        >
          <i
            class="fas fa-hdd fa-4x mb-1"
          />
          <h5>
            /dev/sdb
          </h5>
        </div>
        <div
          class="col-md-6"
        >
          <div
            class="card-body"
          >
            <div
              class="table-responsive"
            >
              <table
                class="table table-borderless table-hover"
              >
                <tbody>
                  <tr>
                    <th
                      scope="row"
                    >
                      Device
                    </th>
                    <td>
                      /dev/sdb
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      UUID
                    </th>
                    <td>
                      1234-1234-1234-1234-1234
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      RAID
                    </th>
                    <td>
                      single
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
    <h2>
      Prepare Drives
    </h2>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="1"
      >
        RAID
      </label>
      <select
        class="form-select"
        id="1"
      >
        <option
          value="custom"
        >
          Not specified
        </option>
        <option
          value="single"
        >
          JBOD
        </option>
        <option
          value="raid1"
        >
          RAID1
        </option>
      </select>
      <div
        class="form-text"
      >
        <small>
          Not specified: 
Don't change the RAID level, keeps everything set the way it was. Useful if you have a really custom setup of your RAID
we don't support and want to just add/remove some disks.

        </small>
      </div>
    </div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th
              scope="col"
            >
              Device
            </th>
            <th
              scope="col"
            >
              Description
            </th>
            <th
              scope="col"
            >
              Filesystem
            </th>
            <th
              scope="col"
            >
              UUID
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class=""
            >
              <div
                class="mb-0 form-check"
              >
                <input
                  class="form-check-input"
                  id="2"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="2"
                >
                  sdb
                </label>
              </div>
            </td>
            <td>
              USB DISK 2.0 (7.2 GiB)
            </td>
            <td />
            <td>
              1111
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              <div
                class="mb-0 form-check"
              >
                <input
                  class="form-check-input"
                  id="3"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="3"
                >
                  sdb1
                </label>
              </div>
            </td>
            <td>
              USB DISK (7.2 GiB)
            </td>
            <td>
              ext4
            </td>
            <td>
              3333
            </td>
          </tr>
          <tr>
            <td
              class=""
            >
              <div
                class="mb-0 form-check"
              >
                <input
                  class="form-check-input"
                  id="4"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="4"
                >
                  sdc
                </label>
              </div>
            </td>
            <td>
              USB DISK 2.0 (3.6 GiB)
            </td>
            <td>
              btrfs
            </td>
            <td>
              4444
            </td>
          </tr>
          <tr>
            <td
              class="text-center"
            >
              <div
                class="mb-0 form-check"
              >
                <input
                  class="form-check-input"
                  id="5"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="5"
                >
                  sdc1
                </label>
              </div>
            </td>
            <td>
              USB DISK (3.6 GiB)
            </td>
            <td>
              vfat
            </td>
            <td>
              2222
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="text-end"
    >
      <button
        class="btn btn-primary mb-2 mb-md-0 mb-lg-3 ml-auto col-12 col-md-4 col-lg-2 d-inline-flex justify-content-center align-items-center"
        data-testid="format-set-btn"
        disabled=""
        type="button"
      >
        <span>
          Format & Set
        </span>
      </button>
    </div>
    <h2>
      Drive Selection
    </h2>
    <p>
      Here you can select which one of the already prepared drives should be
used as external storage. If you already have a device that you used
with the Turris router before, you can set it to be used as external
storage here. Alternatively, if you are migrating your drive to another
Turris device, you can disable external storage here. In such a case,
all your data will remain on the external drive that you can connect
elsewhere. Your /srv will use internal storage again but no data will
be migrated. That means that services like MySQL will end up with empty
databases and LXC virtual machines will be unavailable.
    </p>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th
              scope="col"
            >
              Devices
            </th>
            <th
              scope="col"
            >
              UUID
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <div
                class="mb-0"
              >
                <input
                  class="form-check-input me-2"
                  id="4444"
                  name="uuids"
                  type="radio"
                  value="4444"
                />
                <label
                  class="form-check-label"
                  for="4444"
                >
                  sdc
                </label>
              </div>
            </td>
            <td>
              4444
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="d-grid gap-2 d-md-flex justify-content-md-end"
    >
      <button
        class="btn btn-primary col-sm-12 col-md-4 col-lg-2 mr-md-2 mb-0 mb-md-0 d-inline-flex justify-content-center align-items-center"
        data-testid="disable-external-storage-btn"
        type="button"
      >
        <span>
          Disable external storage
        </span>
      </button>
      <button
        class="btn btn-primary col-sm-12 col-md-4 col-lg-2 d-inline-flex justify-content-center align-items-center"
        data-testid="use-drive-btn"
        type="button"
      >
        <span>
          Use drive
        </span>
      </button>
    </div>
  </div>
</div>
`;
